dialectic-compiler/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+                for (i, c) in cs.iter().enumerate() {
+                    write!(f, "{}", c)?;
+                    if i + 1 < count {
+                        write!(f, ", ")?;
+                    }
+                }
+                if count == 1 {
+                    write!(f, ",")?;
+                }
+                write!(f, ")>")?;
+            }
+            Continue(n) => {
+                write!(f, "Continue")?;
+                if *n > 0 {
+                    write!(f, "<_{}>", n)?;
+                }
+            }
+            Type(s) => write!(f, "{}", s.to_token_stream())?,
+            Error => write!(f, "!")?,
+        }
+        Ok(())
+    }
+}
+
+impl ToTokens for Target {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        use Target::*;
+
+        lazy_static! {
+            static ref CRATE_NAME: String = proc_macro_crate::crate_name("dialectic")
+                .unwrap_or_else(|_| "dialectic".to_owned());
+        }
+
+        let c = Ident::new(&**CRATE_NAME, Span::call_site());
+
+        match self {
+            Done => quote! { #c::types::Done }.to_tokens(tokens),
+            Recv(t, s) => quote! { #c::types::Recv<#t, #s> }.to_tokens(tokens),
+            Send(t, s) => quote! { #c::types::Send<#t, #s> }.to_tokens(tokens),
+            Loop(s) => quote! { #c::types::Loop<#s> }.to_tokens(tokens),
+            Split(s, p) => quote! { #c::types::Split<#s, #p> }.to_tokens(tokens),
+            Call(s, p) => quote! { #c::types::Call<#s, #p> }.to_tokens(tokens),
+            Then(s, p) => quote! { <#s as #c::types::Then<#p>>::Combined }.to_tokens(tokens),
+            Choose(cs) => quote! { #c::types::Choose<(#(#cs,)*)> }.to_tokens(tokens),
+            Offer(cs) => quote! { #c::types::Offer<(#(#cs,)*)> }.to_tokens(tokens),
+            Continue(n) => {
+                if *n > 0 {
+                    (quote! { #c::types::Continue< }).to_tokens(tokens);
+                    (0..*n).for_each(|_| (quote! { #c::types::S< }).to_tokens(tokens));
+                    (quote! { #c::types::Z }).to_tokens(tokens);
+                    (0..=*n).for_each(|_| (quote! { > }).to_tokens(tokens));
+                } else {
+                    (quote! { #c::types::Continue }).to_tokens(tokens);
+                }
+            }
+            Type(s) => quote! { #s }.to_tokens(tokens),
+            Error => quote! { ! }.to_tokens(tokens),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    impl Cfg {
+        fn send(&mut self, ty: &str) -> Index {
+            self.singleton(Ir::Send(syn::parse_str(ty).unwrap()))
+        }
+
+        fn recv(&mut self, ty: &str) -> Index {
+            self.singleton(Ir::Recv(syn::parse_str(ty).unwrap()))
+        }
+
+        fn type_(&mut self, ty: &str) -> Index {
+            self.singleton(Ir::Type(syn::parse_str(ty).unwrap()))
+        }
+    }
+
+    #[test]
+    fn tally_client_expr_direct_subst() {
+        let mut soup = Cfg::new();
```

Comment:
We're not calling it `soup` anymore in the source; let's update the tests to match the nomenclature? :)